design/ceph/ceph-pull-realm.md

```diff
@@ -0,0 +1,94 @@
+# Rook Ceph Realm
+
+## Prerequisites
+
+A Rook Ceph cluster.
+
+
+A Rook Ceph cluster. Ideally a ceph-realm resource would have been started up already.
+
+## Ceph Multisite Overview
+
+Multisite is a feature of Ceph that allows object stores to replicate its data over multiple Ceph clusters. 
+
+Multisite also allows object stores to be independent and isloated from other object stores in a cluster.
+
+For more information on Ceph multsite, please read the [ceph-realm](/design/ceph/ceph-realm.md) design document or [Ceph's documentation](http://docs.ceph.com/docs/master/radosgw/multisite/) on the feature.
+
+## Ceph Pull Realm Walkthrough
+
+### Pulling the Realm from another Ceph Cluster
+
+#### Config
+
+When the storage admin is ready to sync data from a Ceph cluster with multisite set up (primary cluster) to another Rook Ceph cluster (pulling cluster), the pulling cluster needs to have information about the primary cluster's realm and zone group.
+
+To do this, the storage admin needs to create a ceph-pull-realm resource on the pulling cluster with the same name as the realm from the primary cluster.
+
+The endpoint in the `pullEndpoint` section is an endpoint of an object-store in the master zone of the realm. 
+
+This endpoint must be resolvable from the pulling cluster.
+
+This example `ceph-pull-realm.yaml`, pulls `my-realm` from a primary Ceph cluster with endpoint http://my-realm-zone-b-objectstore-1:80 in its master zone.
+```yaml
+apiVersion: ceph.rook.io/v1alpha1
+kind: CephPullRealm
```

Comment:
> I'm not sure this belongs here. Once a realm has been created (either via a pull or via a fresh realm create), it should be able to stand on its own. In addition, it seems a bit counter intuitive to have one k8s cluster with a realm without pullEndpoint with the other replicating k8s cluster with a realm with the pullEndpoint when logically they should be the same object.

> I'd instead consider creating a different resource type called something like CephRealmPull with the ergonomics of a short-lived resource (i.e. Job, Pod, etc). Once its been "processed" successfully, it is no longer needed and could potentially be garbage collected. Also one could put the access key and secret key on this resource. (or a secretRef might have better semantics).

@travisn what happens when the secondary zone becomes a master zone? Or what if a pull is done in the toolbox that is different to what is described on the resource? The idea here is that by putting the pull endpoint on a separate short-lived resource you wouldn't have to worry about keeping track of updating the realm when/if the pullendpoint changes or is no longer relevant. I agree that a verb is a possible indication of something being off. However, it should still be possible to declaratively describe a short-lived task such as a realm pull. It might help to move `Pull` to the rear of the resource name (`CephRealmPull`).

Reply:
My understanding of the realms is that the CRs will only control them at initialization and will not perform any configuration or reconciliation on their desired state after that. If the realm is to be created, the realm CR would have a certain set of properties. If the realm is expected to already exist and just needs to be pulled, it would have a different set of properties including the pull endpoint. But it can be the same CR type. 

In the latest design, the CR does not include whether the realm is intended to be master because that can change more dynamically, such as through the toolbox.

Reply:
it solves the issue where the two ceph clusters belong to the same k8s cluster.
in this case (unless different namesapces are used), there will be a conflict between the two realm definitions. if we use the PullRealm resource, the new Realm CRD will be identical to the existing one, and creating it would be a no-op

Reply:
@travisn: The reason I like having the pull realm, is that is automatically creates the zone group and realm CRs based on the configuration from the other cluster. Though the realm has almost no configuration, the zone group will have quite a bit of configuration. I see the automatic creation of that CR as a way to ensure both clusters have the same multisite CR model (once the CephPullRealm is complete). If that's not reason enough to have a separate CR, then I think we should go back to having just the ceph-realm CR.

@muff1nman: As Travis mentioned, this PR is only for initialization and nothing past that. There are quite a few design considerations to take into account when thinking about how day 2 operations are going to take place, and I'd be happy to hear your thoughts on that when we get to that point.

@yuvalif: I asked @travisn about this situation and he told me there can only be one rook cluster per k8s namespace. The situation you described is not practical or possible.

Reply:
> @travisn: The reason I like having the pull realm, is that is automatically creates the zone group and realm CRs based on the configuration from the other cluster. Though the realm has almost no configuration, the zone group will have quite a bit of configuration. I see the automatic creation of that CR as a way to ensure both clusters have the same multisite CR model (once the CephPullRealm is complete). If that's not reason enough to have a separate CR, then I think we should go back to having just the ceph-realm CR.

@alimaredia To clarify terminology, I don't think the CRs are being automatically created. You're referring to automatically pulling in the realm and zone groups to be created at the Ceph level, right? The CR should only refer to the declarative manifests. Instead of defining a new CR type for pulling the realm, I'm proposing to have the `pull` properties in the same realm CRD type.

Reply:
> @muff1nman: As Travis mentioned, this PR is only for initialization and nothing past that. There are quite a few design considerations to take into account when thinking about how day 2 operations are going to take place, and I'd be happy to hear your thoughts on that when we get to that point.

I understand the desire here to get an MVP that solves the day 1 story, but I think its important to have a design that won't need to get discarded for the day 2..n story. Specifically, I'd like to think of an implementation where the following is true:
1. CRD get via kubectl provides the most necessary information to reason about realms, zonegroups, zones, their sync status across different k8s instances
2. CRD edit/apply via kubectl can solve nearly all of the day to day operations that are needed to maintain a multisite setup. The toolbox is used in the rare case of debugging/repair.
3. removal of zones, zonegroups, realms is possible via deleting the CRD

I believe that in order to enable the above, it is important and necessary to stay as close to ceph model as possible so that *bidirectional* syncing can be done by the operator.

> My understanding of the realms is that the CRs will only control them at initialization and will not perform any configuration or reconciliation on their desired state after that

>  I don't think the CRs are being automatically created. 

I think they should be. After a `CephRealmPull` the operator should create all of the CR to be in sync with the information in ceph and therefore in sync with the source/master k8s resources. This is important because it enables modifications/deletions (at a later time down the road). It also enables a straightforward path to provide the status information needed for point 1.

Alternatively, if this initial implementation is truly only to be designed for the happy path of a single realm pull and nothing more, I would go back to the initial design (and maybe a step further) with the realm pull information on the `CephObjectStore`. I would even go further and put it on an annotation. Doing it as an annotation would:
- allow iterating faster without having to formalize types
- be more clear to the user that the value is immutable / only relevant at creation time (there are no guarantees/expectations about CREATE/UPDATE/DELETE on an annotation)
- no status information is expected with an annotation

Reply:
> > @muff1nman: As Travis mentioned, this PR is only for initialization and nothing past that. There are quite a few design considerations to take into account when thinking about how day 2 operations are going to take place, and I'd be happy to hear your thoughts on that when we get to that point.
> 
> I understand the desire here to get an MVP that solves the day 1 story, but I think its important to have a design that won't need to get discarded for the day 2..n story. Specifically, I'd like to think of an implementation where the following is true:
> 
> 1. CRD get via kubectl provides the most necessary information to reason about realms, zonegroups, zones, their sync status across different k8s instances
> 2. CRD edit/apply via kubectl can solve nearly all of the day to day operations that are needed to maintain a multisite setup. The toolbox is used in the rare case of debugging/repair.
> 3. removal of zones, zonegroups, realms is possible via deleting the CRD
> 
> I believe that in order to enable the above, it is important and necessary to stay as close to ceph model as possible so that _bidirectional_ syncing can be done by the operator.
> 
> > My understanding of the realms is that the CRs will only control them at initialization and will not perform any configuration or reconciliation on their desired state after that
> 
> > I don't think the CRs are being automatically created.
> 
> I think they should be. After a `CephRealmPull` the operator should create all of the CR to be in sync with the information in ceph and therefore in sync with the source/master k8s resources. This is important because it enables modifications/deletions (at a later time down the road). It also enables a straightforward path to provide the status information needed for point 1.
> 

agree with this. Don't think it makes the solution much complex, and would allow updates and deletion to be easily implemented later on.

> Alternatively, if this initial implementation is truly only to be designed for the happy path of a single realm pull and nothing more, I would go back to the initial design (and maybe a step further) with the realm pull information on the `CephObjectStore`. I would even go further and put it on an annotation. Doing it as an annotation would:
> 
> * allow iterating faster without having to formalize types
> * be more clear to the user that the value is immutable / only relevant at creation time (there are no guarantees/expectations about CREATE/UPDATE/DELETE on an annotation)
> * no status information is expected with an annotation